hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMFailoverProxyProvider.java

```diff
@@ -261,9 +263,9 @@ public RetryAction shouldRetry(Exception exception, int retries,
             //  address of the suggested leader along with the nodeID.
             //  Failing over just based on nodeID is not very robust.
 
-            // OMFailoverProxyProvider#performFailover() is a dummy call and
-            // does not perform any failover. Failover manually to the next OM.
-            performFailoverToNextProxy();
+            // Prepare the next OM to be tried. This will help with calculation
+            // of the wait times needed get creating the retryAction.
+            selectNextOmProxy();
```

Comment:
With multiple threads, in the case when the `nextOmProxyIndex` is set by one thread calling `shouldRetry` and another thread executes `shouldRetry` prior to the `performFailover` (setting of the` currentOmNodeId`), the _nextOmProxyIndex_ could be incremented before the OmProxy is tried.  Can this be a problem?  Should we not check if the _nextOmProxyIndex_ == `currentOmProxyIndex` before trying the next proxy (calling `selectNextOmProxy`), some sort of synchronized check?
If `nextOmProxyIndex` != `currentOmProxyIndex `(unsynchronized), then it implies another thread is in failover and has just executed `shouldRetry` prior to calling `performFailover`.  We need to try the current `nextOmProxyIndex` before trying other.

Reply:
Agree.

Even though performFailover will be called only once by concurrent failed invocations (guaranteed by RetryInvocationHandler), this code does not guarantee that the failover will occur to the next proxy. The failover could as well happen to the same proxy if we do not ensure that selectNextOmProxy is effective only once for concurrent failovers (same as before). 
Having nextOmProxyId == currentOmProxyId check would help here. 

The reason for not using the RetryInvocationHandler#performFailover() to do the actual failover was that with the base method we could not failover to the suggested leader OM (which we are anyway not doing currently but it would be good to add it back). 

I am thinking if we should update the attemptedOMs list (used to calculate waitTime) from performFailover instead of from selectNextOmProxy.

Reply:
Agreed. I will make a slightly different change and update.